Eliminate read past end of array in percentile_weightnening

wphicks opened the pull request:
Because the existing percentile_weightnening kernel reads past the end of the input array, the 100th quantile returned by `cupy.percentile` can (incorrectly) be a NaN if that extra read returns a NaN as described in #4451. This PR ensures that the kernel does not read past the end of the array by passing in the size of the array and ensuring that the index is at most `size - 1`.

While the fix is straightforward, the associated test is somewhat trickier. The unit test provided as part of this PR allocates a large array of NaNs, deletes that array, and then calls `cupy.percentile` in order to maximize the likelihood that arrays newly-allocated during the computation will have a NaN in device memory just past their end.

I'm currently leaving this PR as WIP because I would appreciate feedback on the included test. While this test consistently fails for me without the described fix, I don't have a good theoretical justification that this would correctly fail every time. More specifically, the size of the allocated array of NaNs is somewhat arbitrary. 1000 produced consistent failures on my system, but 100 did not.

At a higher level, is this unit test worth having? It demonstrates the very specific issue in the initial `percentile_weightnening` kernel, but `cuda-memcheck` is probably better suited to finding this class of errors than trying to guard against them in tests.

Resolve #4451

takagi commented:
pfnCI, test this please.

takagi commented:
Thanks for your quick fix! As you mark this WIP, the unit test implicitly depends on the behavior of CuPy's memory pool facility. I think It is acceptable, but please let me think of a clearer way to do that just a while.

wphicks commented:
Happy to help! Thank you for giving the unit test some thought. I've also asked the rest of my team to consider other possibilities for that test, and I'm sure they'll post here if they think of a better alternative.

takagi commented:
The following code should always reproduce the issue. Here, I hook CuPy's default allocator to view the memory block it allocates as float32, filling its contents with `cupy.nan` before it returns the memory pointer to the block. Please not that CuPy's allocator takes a memory block of a multiple of 512 bytes, so the ndarray `ap` [here](https://github.com/cupy/cupy/blob/bcbaf919c49942a146597f281c4a15f2836b08d8/cupy/_statistics/order.py#L200) is always followed by another 4 bytes filled with NaN out of its shape. Would you check if it works?

```Python
import cupy
from cupy import cuda

a = cupy.arange(5, dtype='f')
q = cupy.array([0, 50, 100], dtype='f')

orig = cuda.get_allocator()
def alloc(size):
    memptr = orig(size)
    base_size = memptr.mem.size
    assert base_size == 512
    item_size = cupy.float32().itemsize
    shape = base_size // item_size,
    x = cupy.ndarray(memptr=memptr, shape=shape, dtype='f')
    x.fill(cupy.nan)
    return memptr

cuda.set_allocator(alloc)
try:
    cupy.percentile(a, q)
finally:
    cupy.cuda.set_allocator(orig)
```

wphicks commented:
Brilliant! That works perfectly. I've just updated the unit test to make use of this. I've modified it very slightly to generalize across the various dtypes we use in the test; let me know if you see any issues with those tweaks.

I'm going to go ahead and pull off the WIP since your updated test seems entirely robust. Thanks very much for the help!

takagi commented on tests/cupy_tests/statistics_tests/test_order.py:
```diff
@@ -6,7 +6,7 @@
 import cupy
 import cupy.core._accelerator as _acc
-from cupy import testing
+from cupy import cuda, testing
```
Would you split this line to import a module per line to follow our guideline that requires it in general?
```suggestion
from cupy import cuda
from cupy import testing
```

takagi commented on tests/cupy_tests/statistics_tests/test_order.py:
```diff
@@ -117,6 +117,35 @@ def test_percentile_unexpected_interpolation(self, dtype):
             with pytest.raises(ValueError):
                 xp.percentile(a, q, axis=-1, interpolation='deadbeef')
 
+    @testing.for_float_dtypes()
```
Would you add a comment to refer to GitHub to show why this test is provided?
```suggestion
    # See gh-4453
    @testing.for_float_dtypes()
```

takagi reviewed (changes requested):
Thanks for checking it and updating the unit test! The change is clearly fine. Would you make some minor changes as commented? After that, I will kick the CI again and merge this change to the master.

wphicks commented:
Thanks very much! Updated accordingly.

takagi commented:
pfnCI, test this please.

takagi commented:
LGTM, thanks! The change will be merged into the master after the CI passes.
